File: UrbanHomeServices/app/src/main/java/com/huawei/hms/urbanhomeservices/java/activities/SplashActivity.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2020-2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.hms.urbanhomeservices.java.activities;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.huawei.hms.urbanhomeservices.R;
import com.huawei.hms.urbanhomeservices.java.clouddb.LoginInfo;
import com.huawei.hms.urbanhomeservices.java.utils.AppConstants;
import com.huawei.hms.urbanhomeservices.java.utils.AppPreferences;
import com.huawei.hms.urbanhomeservices.java.utils.Utils;

/**
 * This activity basically helps to select user type :
 * 1 : Consumer type
 * 2 : Service provider type.
 *
 * @author: Huawei
 * @since : 20-01-2021
 */

public class SplashActivity extends AppCompatActivity {

    private static final String TAG = SplashActivity.class.getSimpleName();
    private LoginInfo loginInfo;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_splash);
        locationPermission();
        checkUserLoginType();
        loginInfo = new LoginInfo();
        String[] items = getResources().getStringArray(R.array.select_user_type);
        ArrayAdapter adapter = new ArrayAdapter(this, android.R.layout.simple_spinner_item, items);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        Spinner spnLogin = findViewById(R.id.spnLogin);
        spnLogin.setAdapter(adapter);
        Button btnNext = findViewById(R.id.btnnext);
        btnNext.setOnClickListener(v -> {
            if (spnLogin.getSelectedItemId() == AppConstants.INTIAL_VALUE) {
                Utils.showToast(v.getContext(), getString(R.string.select_type));
            }
            checkLogin(spnLogin);
        });
    }

    /**
     * Check if the user is already signed in based on user type and navigate to particular activity
     */
    private void checkUserLoginType() {
        if (AppPreferences.isLogin() && AppPreferences.getUserType().equals(getResources().getString(R.string.select_consumer_type))) {
            Intent mainIntent = new Intent(SplashActivity.this, MainActivity.class);
            startActivity(mainIntent);
            finish();
        } else if (AppPreferences.isLogin() && AppPreferences.getUserType().equals(getResources().getString(R.string.select_provider_type))) {
            Intent intent = new Intent(SplashActivity.this, AddServiceActivity.class);
            intent.putExtra(AppConstants.LOGIN_USER_TYPE, getResources().getString(R.string.select_provider_type));
            startActivity(intent);
            finish();
        }
    }

    /**
     * Select the user type from spinner
     *
     * @param spnLogin user type drop down.
     */
    private void checkLogin(Spinner spnLogin) {
        if (spnLogin.getSelectedItemId() == AppConstants.INITIAL_VALUE_ONE) {
            if (AppPreferences.isLogin() && AppPreferences.getUserType().equals(getResources().getString(R.string.select_consumer_type))) {
                Intent mainIntent = new Intent(SplashActivity.this, MainActivity.class);
                AppPreferences.setIsLogin(true);
                AppPreferences.setUserType(getResources().getString(R.string.select_consumer_type));
                mainIntent.putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.getSelectedItem().toString());
                mainIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(mainIntent);
                finish();
            } else {
                Intent loginIntent = new Intent(SplashActivity.this, LoginActivity.class);
                loginIntent.putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.getSelectedItem().toString());
                startActivity(loginIntent);
            }
        } else if (spnLogin.getSelectedItemId() == AppConstants.INITIAL_VALUE_TWO) {
            if (AppPreferences.isLogin() && AppPreferences.getUserType().equals(getResources().getString(R.string.select_provider_type))) {
                Intent intent = new Intent(SplashActivity.this, AddServiceActivity.class);
                AppPreferences.setIsLogin(true);
                AppPreferences.setUserType(getResources().getString(R.string.select_provider_type));
                intent.putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.getSelectedItem().toString());
                intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(intent);
                finish();
            } else {
                Intent loginIntent = new Intent(SplashActivity.this, LoginActivity.class);
                loginIntent.putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.getSelectedItem().toString());
                startActivity(loginIntent);
            }
        }
    }

    /**
     * Asking for location permission
     * which is used to fetch current location on home Fragment
     */
    private void locationPermission() {
        if (ContextCompat.checkSelfPermission(SplashActivity.this,
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(SplashActivity.this,
                    Manifest.permission.ACCESS_FINE_LOCATION)) {
                ActivityCompat.requestPermissions(SplashActivity.this,
                        new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
            } else {
                ActivityCompat.requestPermissions(SplashActivity.this,
                        new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        switch (requestCode) {
            case 1: {
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    if (ContextCompat.checkSelfPermission(SplashActivity.this,
                            Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                        Utils.showToast(this, "Permission Granted");
                    }
                } else {
                    Utils.showToast(this, "Permission Denied");
                }
            }
        }
    }
}

File: UrbanHomeServices/app/src/main/java/com/huawei/hms/urbanhomeservices/kotlin/activities/SplashActivity.kt
/*
 *
 *  * Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *  *
 *  *    Licensed under the Apache License, Version 2.0 (the "License");
 *  *    you may not use this file except in compliance with the License.
 *  *    You may obtain a copy of the License at
 *  *
 *  *      http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  *    Unless required by applicable law or agreed to in writing, software
 *  *    distributed under the License is distributed on an "AS IS" BASIS,
 *  *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *    See the License for the specific language governing permissions and
 *  *    limitations under the License.
 *
 */
@file:Suppress("MaxLineLength","NewLineAtEndOfFile")
package com.huawei.hms.urbanhomeservices.kotlin.activities

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.widget.ArrayAdapter
import android.widget.Spinner
import androidx.appcompat.app.AppCompatActivity
import com.huawei.hms.urbanhomeservices.R
import com.huawei.hms.urbanhomeservices.databinding.ActivitySplashBinding
import com.huawei.hms.urbanhomeservices.databinding.ManageServiceActivityBinding
import com.huawei.hms.urbanhomeservices.kotlin.clouddb.LoginInfo
import com.huawei.hms.urbanhomeservices.kotlin.utils.AppConstants
import com.huawei.hms.urbanhomeservices.kotlin.utils.AppPreferences
import com.huawei.hms.urbanhomeservices.kotlin.utils.Utils


/**
 * This activity basically helps to select user type :
 * 1 : Consumer type
 * 2 : Service provider type.
 *
 * @author: Huawei
 * @since : 20-01-2021
 */
class SplashActivity : AppCompatActivity() {
    private val tag = "SplashActivity"
    private var loginInfo: LoginInfo? = null

    private lateinit var splashBinding: ActivitySplashBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        splashBinding = ActivitySplashBinding.inflate(layoutInflater)
        val view = splashBinding.root
        setContentView(view)
        locationPermission()
        checkUserLoginType()
        loginInfo = LoginInfo()
        val items: Array<String> = resources.getStringArray(R.array.select_user_type)
        val adapter: ArrayAdapter<String> = ArrayAdapter(
                this,
                android.R.layout.simple_spinner_item, items
        ).apply {
            setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }
        splashBinding.spnLogin.adapter = adapter
        splashBinding.btnnext.setOnClickListener {
            splashBinding.spnLogin.selectedItemId.toString()
            if (splashBinding.spnLogin.selectedItemId.toString() == AppConstants.INTIAL_VALUE.toString()) {
                Utils.showToast(this@SplashActivity, getString(R.string.select_type))
                return@setOnClickListener
            }
            checkLogin(splashBinding.spnLogin)
        }
    }

    /**
     * Check if the user is already signed in based on user type and navigate to particular activity
     */
    private fun checkUserLoginType() {
        if (AppPreferences.isLogin && AppPreferences.userType.equals(resources.getString(R.string.select_consumer_type))) {
            startActivity(Intent(this@SplashActivity, MainActivity::class.java))
            finish()
        } else if (AppPreferences.isLogin && AppPreferences.userType.equals(resources.getString(R.string.select_provider_type))) {
            val intent = Intent(
                    this@SplashActivity,
                    AddServiceActivity::class.java
            )
            intent.putExtra(
                    AppConstants.LOGIN_USER_TYPE,
                    resources.getString(R.string.select_provider_type)
            )
            startActivity(intent)
            finish()
        }
    }
    @Suppress("ComplexMethod")
    /**
     * Select the user type from spinner
     *
     * @param spnLogin user type drop down.
     */
    private fun checkLogin(spnLogin: Spinner) {
        if (spnLogin.selectedItemId.toString() == AppConstants.INITIAL_VALUE_ONE.toString()) {
            if (AppPreferences.isLogin && AppPreferences.userType.equals(resources.getString(R.string.select_consumer_type))) {
                val intent = Intent(this@SplashActivity, MainActivity::class.java)
                AppPreferences.apply {
                    isLogin = true
                    userType = resources.getString(R.string.select_consumer_type)
                }
                intent.apply {
                    putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.selectedItem.toString())
                    flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                }
                startActivity(intent)
                finish()
            } else {
                val intent = Intent(this@SplashActivity, LoginActivity::class.java)
                intent.putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.selectedItem.toString())
                startActivity(intent)
            }
        } else if (spnLogin.selectedItemId.toString() == AppConstants.INITIAL_VALUE_TWO.toString()) {
            if (AppPreferences.isLogin && AppPreferences.userType.equals(resources.getString(R.string.select_provider_type))) {
                val intent = Intent(this@SplashActivity, AddServiceActivity::class.java)
                AppPreferences.apply {
                    isLogin = true
                    userType = resources.getString(R.string.select_provider_type)

                }
                intent.apply {
                    putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.selectedItem.toString())
                    flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                }
                startActivity(intent)
                finish()
            } else {
                val intent = Intent(this@SplashActivity, LoginActivity::class.java)
                intent.putExtra(AppConstants.LOGIN_USER_TYPE, spnLogin.selectedItem.toString())
                startActivity(intent)
            }
        }
    }

    /**
     * Asking for location permission
     * which is used to fetch current location on home Fragment
     */
    private fun locationPermission() {
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                    && checkSelfPermission(Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED
            ) {
                val strings = arrayOf(
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION
                )
                requestPermissions(strings, AppConstants.INITIAL_VALUE_ONE)
            }
        } else {
            if (checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                    && checkSelfPermission(
                            Manifest.permission.ACCESS_COARSE_LOCATION
                    ) != PackageManager.PERMISSION_GRANTED
                    && checkSelfPermission(AppConstants.LOGIN_LOCATION_PERMISSION) != PackageManager.PERMISSION_GRANTED
            ) {
                val strings = arrayOf(
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION,
                        AppConstants.LOGIN_LOCATION_PERMISSION
                )
                requestPermissions(strings, AppConstants.INITIAL_VALUE_TWO)
            }
        }
    }
    @Suppress("ComplexCondition")
    /**
     * Verifies location permission is granted or not
     */
    override fun onRequestPermissionsResult(
            requestCode: Int,
            permissions: Array<String?>,
            grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == AppConstants.INITIAL_VALUE_ONE) {
            if (grantResults.size > AppConstants.INITIAL_VALUE_ONE &&
                    grantResults[AppConstants.INTIAL_VALUE] == PackageManager.PERMISSION_GRANTED &&
                    grantResults[AppConstants.INITIAL_VALUE_ONE] == PackageManager.PERMISSION_GRANTED
            ) {
                Log.i(tag, "onRequestPermissionsResult: apply LOCATION PERMISSION successful")
            } else {
                Log.i(tag, "onRequestPermissionsResult: apply LOCATION PERMISSION  failed")
            }
        }
        if (requestCode == AppConstants.INITIAL_VALUE_TWO) {
            if (grantResults.size > AppConstants.INITIAL_VALUE_TWO
                    && grantResults[AppConstants.INITIAL_VALUE_TWO] == PackageManager.PERMISSION_GRANTED
                    && grantResults[AppConstants.INTIAL_VALUE] == PackageManager.PERMISSION_GRANTED
                    && grantResults[AppConstants.INITIAL_VALUE_ONE] == PackageManager.PERMISSION_GRANTED) {
                Log.i(
                        tag,
                        "onRequestPermissionsResult: apply ACCESS_BACKGROUND_LOCATION successful"
                )
            } else {
                Log.i(tag, "onRequestPermissionsResult: apply ACCESS_BACKGROUND_LOCATION  failed")
            }
        }
    }
}
